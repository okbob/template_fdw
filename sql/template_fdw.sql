CREATE EXTENSION IF NOT EXISTS template_fdw;

CREATE SERVER template FOREIGN DATA WRAPPER template_fdw;

CREATE FOREIGN TABLE foo(a int, b int) SERVER template;

-- should fail
INSERT INTO foo VALUES(10,20);
SELECT * FROM foo;

-- should work
CREATE TEMP TABLE foo(LIKE foo INCLUDING ALL);
INSERT INTO foo VALUES(10,20);
SELECT * FROM foo;

