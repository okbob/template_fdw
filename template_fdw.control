# template_fdw extension
comment = 'a foreign data wrapper for template storage'
default_version = '0.0'
module_pathname = '$libdir/template_fdw'
relocatable = true
