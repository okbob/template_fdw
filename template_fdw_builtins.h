
#ifndef TEMPLATE_FDW_BUILTINS
#define TEMPLATE_FDW_BUILTINS

#ifndef PGDLLEXPORT
#ifdef _MSC_VER
#define PGDLLEXPORT	__declspec(dllexport)

/*
 * PG_MODULE_MAGIC and PG_FUNCTION_INFO_V1 macros are broken for MSVC.
 * So, we redefine them.
 */

#undef PG_MODULE_MAGIC
#define PG_MODULE_MAGIC \
extern PGDLLEXPORT const Pg_magic_struct *PG_MAGIC_FUNCTION_NAME(void); \
const Pg_magic_struct * \
PG_MAGIC_FUNCTION_NAME(void) \
{ \
	static const Pg_magic_struct Pg_magic_data = PG_MODULE_MAGIC_DATA; \
	return &Pg_magic_data; \
} \
extern int no_such_variable

#undef PG_FUNCTION_INFO_V1
#define PG_FUNCTION_INFO_V1(funcname) \
extern PGDLLEXPORT const Pg_finfo_record * CppConcat(pg_finfo_,funcname)(void); \
const Pg_finfo_record * \
CppConcat(pg_finfo_,funcname) (void) \
{ \
	static const Pg_finfo_record my_finfo = { 1 }; \
	return &my_finfo; \
} \
extern int no_such_variable

#else
#define PGDLLEXPORT	PGDLLIMPORT
#endif
#endif

/*
 * Interface
 *
 */
extern PGDLLEXPORT Datum template_fdw_validator(PG_FUNCTION_ARGS);
extern PGDLLEXPORT Datum template_fdw_handler(PG_FUNCTION_ARGS);

#endif
