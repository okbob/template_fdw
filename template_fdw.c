/*-------------------------------------------------------------------------
 *
 * template_fdw.c
 *
 *			  a template foreign data wrapper
 *
 * by Pavel Stehule 2013, 2014, 2015
 *
 *-------------------------------------------------------------------------
 *
 * Notes:
 *
 *   This wrapper doesn't allow any DML or SELECT operation.
 *
 */
#include "postgres.h"

#include "funcapi.h"
#include "miscadmin.h"


#include "access/reloptions.h"
#include "foreign/fdwapi.h"
#include "foreign/foreign.h"
#include "optimizer/pathnode.h"
#include "optimizer/planmain.h"
#include "optimizer/restrictinfo.h"
#include "utils/builtins.h"
#include "utils/lsyscache.h"
#include "utils/rel.h"

#include "template_fdw_builtins.h"

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

/* callback functions */
static void templateGetForeignRelSize(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid);
static void templateGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid);
static ForeignScan *templateGetForeignPlan(PlannerInfo *root, RelOptInfo *baserel,
								 Oid foreigntableid, ForeignPath *best_path,
								 List *tlist, List *scan_clauses);
static void templateBeginForeignScan(ForeignScanState *node, int eflags);
static TupleTableSlot *templateIterateForeignScan(ForeignScanState *node);
static void templateReScanForeignScan(ForeignScanState *node);
static void templateEndForeignScan(ForeignScanState *node);
static void templateAddForeignUpdateTargets(Query *parsetree, RangeTblEntry *target_rte, Relation target_relation);
static List *templatePlanForeignModify(PlannerInfo *root, ModifyTable *plan, Index resultRelation, int subplan_index);
static void templateBeginForeignModify(ModifyTableState *mtstate, ResultRelInfo *rinfo,
										    List *fdw_private,
										    int subplan_index, int eflags);
static TupleTableSlot *templateExecForeignInsert(EState *estate,
								   ResultRelInfo *rinfo, TupleTableSlot *slot,
								   TupleTableSlot *planSlot);
static TupleTableSlot *templateExecForeignUpdate(EState *estate,
								   ResultRelInfo *rinfo, TupleTableSlot *slot,
								   TupleTableSlot *planSlot);
static TupleTableSlot *templateExecForeignDelete(EState *estate,
								   ResultRelInfo *rinfo,
								   TupleTableSlot *slot, TupleTableSlot *planSlot);
static void templateEndForeignModify(EState *estate, ResultRelInfo *rinfo);
static bool templateAnalyzeForeignTable(Relation relation, AcquireSampleRowsFunc *func, BlockNumber *totalpages);

PG_FUNCTION_INFO_V1(template_fdw_handler);
PG_FUNCTION_INFO_V1(template_fdw_validator);

Datum
template_fdw_handler(PG_FUNCTION_ARGS)
{
	FdwRoutine *fdwroutine = makeNode(FdwRoutine);

	/* assign the handlers for the FDW */

	/* these are required */
	fdwroutine->GetForeignRelSize = templateGetForeignRelSize;
	fdwroutine->GetForeignPaths = templateGetForeignPaths;
	fdwroutine->GetForeignPlan = templateGetForeignPlan;
	fdwroutine->BeginForeignScan = templateBeginForeignScan;
	fdwroutine->IterateForeignScan = templateIterateForeignScan;
	fdwroutine->ReScanForeignScan = templateReScanForeignScan;
	fdwroutine->EndForeignScan = templateEndForeignScan;

	/* remainder are optional - use NULL if not required */
	/* support for insert / update / delete */
	fdwroutine->AddForeignUpdateTargets = templateAddForeignUpdateTargets;
	fdwroutine->PlanForeignModify = templatePlanForeignModify;
	fdwroutine->BeginForeignModify = templateBeginForeignModify;
	fdwroutine->ExecForeignInsert = templateExecForeignInsert;
	fdwroutine->ExecForeignUpdate = templateExecForeignUpdate;
	fdwroutine->ExecForeignDelete = templateExecForeignDelete;
	fdwroutine->EndForeignModify = templateEndForeignModify;

	/* support for EXPLAIN */
	fdwroutine->ExplainForeignScan = NULL;
	fdwroutine->ExplainForeignModify = NULL;

	/* support for ANALYSE */
	fdwroutine->AnalyzeForeignTable = templateAnalyzeForeignTable;

	PG_RETURN_POINTER(fdwroutine);
}

Datum
template_fdw_validator(PG_FUNCTION_ARGS)
{
	List	   *options_list = untransformRelOptions(PG_GETARG_DATUM(0));

	if (list_length(options_list) > 0)
		ereport(ERROR,
			(errcode(ERRCODE_FDW_INVALID_OPTION_NAME),
			 errmsg("invalid options"),
			 errhint("template FDW doies not support any options")));

	PG_RETURN_VOID();
}

static void
templateGetForeignRelSize(PlannerInfo *root,
			   RelOptInfo *baserel,
			   Oid foreigntableid)
{
	baserel->fdw_private = NULL;
	baserel->rows = 0;
}

static void
templateGetForeignPaths(PlannerInfo *root,
			 RelOptInfo *baserel,
			 Oid foreigntableid)
{
	add_path(baserel, (Path *)
		 create_foreignscan_path(root, baserel,
							     0, 0, 0,
									 NIL, NULL, NIL));
}

static ForeignScan *
templateGetForeignPlan(PlannerInfo *root,
			RelOptInfo *baserel,
			Oid foreigntableid,
			ForeignPath *best_path,
			List *tlist,
			List *scan_clauses)
{
	Index		scan_relid = 1;

	scan_clauses = extract_actual_clauses(scan_clauses, false);

	/* Create the ForeignScan node */
#if PG_VERSION_NUM >= 90500

	return make_foreignscan(tlist,
					    scan_clauses,  scan_relid,
								        NIL, NIL, NIL);

#else

	return make_foreignscan(tlist,
					    scan_clauses,  scan_relid,
								        NIL, NIL);

#endif

}

static void
templateBeginForeignScan(ForeignScanState *node,
			  int eflags)
{
}


static TupleTableSlot *
templateIterateForeignScan(ForeignScanState *node)
{
	Relation rel = node->ss.ss_currentRelation;
	char *template_name,
			*table_name;

	template_name = quote_qualified_identifier(get_namespace_name(RelationGetNamespace(rel)),
										     RelationGetRelationName(rel));
	table_name = quote_qualified_identifier(NULL, RelationGetRelationName(rel));

	ereport(ERROR,
			(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
			 errmsg("cannot read from table \"%s\"", template_name),
			 errdetail("Table is template."),
			 errhint("Create temp table by statement "
			 "\"CREATE TEMP TABLE %s(LIKE %s INCLUDING ALL);\"", table_name, template_name)));
}


static void
templateReScanForeignScan(ForeignScanState *node)
{
}


static void
templateEndForeignScan(ForeignScanState *node)
{
}


static void
templateAddForeignUpdateTargets(Query *parsetree,
				 RangeTblEntry *target_rte,
				 Relation target_relation)
{
}


static List *
templatePlanForeignModify(PlannerInfo *root,
			   ModifyTable *plan,
			   Index resultRelation,
			   int subplan_index)
{
    return NULL;
}


static void
templateBeginForeignModify(ModifyTableState *mtstate,
			    ResultRelInfo *rinfo,
			    List *fdw_private,
			    int subplan_index,
			    int eflags)
{
}


static TupleTableSlot *
templateExecForeignInsert(EState *estate,
			   ResultRelInfo *rinfo,
			   TupleTableSlot *slot,
			   TupleTableSlot *planSlot)
{

	Relation rel = rinfo->ri_RelationDesc;
	char *template_name,
			*table_name;

	template_name = quote_qualified_identifier(get_namespace_name(RelationGetNamespace(rel)),
										     RelationGetRelationName(rel));
	table_name = quote_qualified_identifier(NULL, RelationGetRelationName(rel));

	ereport(ERROR,
			(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
			 errmsg("cannot insert into to table \"%s\"", template_name),
			 errdetail("Table is template."),
			 errhint("Create temp table by statement "
			 "\"CREATE TEMP TABLE %s(LIKE %s INCLUDING ALL);\"", table_name, template_name)));
}


static TupleTableSlot *
templateExecForeignUpdate(EState *estate,
			   ResultRelInfo *rinfo,
			   TupleTableSlot *slot,
			   TupleTableSlot *planSlot)
{
	Relation rel = rinfo->ri_RelationDesc;
	char *template_name,
			*table_name;

	template_name = quote_qualified_identifier(get_namespace_name(RelationGetNamespace(rel)),
										     RelationGetRelationName(rel));
	table_name = quote_qualified_identifier(NULL, RelationGetRelationName(rel));

	ereport(ERROR,
			(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
			 errmsg("cannot update table \"%s\"", template_name),
			 errdetail("Table is template."),
			 errhint("Create temp table by statement "
			 "\"CREATE TEMP TABLE %s(LIKE %s INCLUDING ALL);\"", table_name, template_name)));
}


static TupleTableSlot *
templateExecForeignDelete(EState *estate,
			   ResultRelInfo *rinfo,
			   TupleTableSlot *slot,
			   TupleTableSlot *planSlot)
{
	Relation rel = rinfo->ri_RelationDesc;
	char *template_name,
			*table_name;

	template_name = quote_qualified_identifier(get_namespace_name(RelationGetNamespace(rel)),
										     RelationGetRelationName(rel));
	table_name = quote_qualified_identifier(NULL, RelationGetRelationName(rel));

	ereport(ERROR,
			(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
			 errmsg("cannot delete from table \"%s\"", template_name),
			 errdetail("Table is template."),
			 errhint("Create temp table by statement "
			 "\"CREATE TEMP TABLE %s(LIKE %s INCLUDING ALL);\"", table_name, template_name)));
}


static void
templateEndForeignModify(EState *estate,
			  ResultRelInfo *rinfo)
{
}

static bool
templateAnalyzeForeignTable(Relation relation,
			     AcquireSampleRowsFunc *func,
			     BlockNumber *totalpages)
{
    return false;
}
