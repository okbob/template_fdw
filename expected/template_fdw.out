CREATE EXTENSION IF NOT EXISTS template_fdw;
CREATE SERVER template FOREIGN DATA WRAPPER template_fdw;
CREATE FOREIGN TABLE foo(a int, b int) SERVER template;
-- should fail
INSERT INTO foo VALUES(10,20);
ERROR:  cannot insert into to table "public.foo"
DETAIL:  Table is template.
HINT:  Create temp table by statement "CREATE TEMP TABLE foo(LIKE public.foo INCLUDING ALL);"
SELECT * FROM foo;
ERROR:  cannot read from table "public.foo"
DETAIL:  Table is template.
HINT:  Create temp table by statement "CREATE TEMP TABLE foo(LIKE public.foo INCLUDING ALL);"
-- should work
CREATE TEMP TABLE foo(LIKE foo INCLUDING ALL);
INSERT INTO foo VALUES(10,20);
SELECT * FROM foo;
 a  | b  
----+----
 10 | 20
(1 row)

